src/cpp/dds/topic/DataWriterImpl.cpp

```diff
@@ -0,0 +1,720 @@
+    {
+        logError(DATA_WRITER, "Problem creating associated Writer");
+    }
+
+    writer_ = writer;
+}
+
+void DataWriterImpl::disable()
+{
+    set_listener(nullptr);
+    if(writer_ != nullptr)
+    {
+        writer_->set_listener(nullptr);
+    }
+}
+
+DataWriterImpl::~DataWriterImpl()
+{
+    delete lifespan_timer_;
+    delete deadline_timer_;
+
+    if(writer_ != nullptr)
+    {
+        logInfo(PUBLISHER, guid().entityId << " in topic: " << type_->getName());
+    }
+
+    RTPSDomain::removeRTPSWriter(writer_);
+    delete user_datawriter_;
+}
+
+bool DataWriterImpl::write(
+        void* data)
+{
+    logInfo(DATA_WRITER, "Writing new data");
+    return create_new_change(ALIVE, data);
+}
+
+bool DataWriterImpl::write(
+        void* data,
+        rtps::WriteParams& params)
+{
+    logInfo(DATA_WRITER, "Writing new data with WriteParams");
+    return create_new_change_with_params(ALIVE, data, params);
+}
+
+bool DataWriterImpl::write(
+            void* data,
+            const rtps::InstanceHandle_t& handle)
+{
+    logInfo(DATA_WRITER, "Writing new data with Handle");
+    WriteParams wparams;
+    return create_new_change_with_params(ALIVE, data, wparams, handle);
+}
+
+bool DataWriterImpl::dispose(
+        void* data,
+        const rtps::InstanceHandle_t& handle)
+{
+    logInfo(DATA_WRITER, "Disposing of data");
+    WriteParams wparams;
+    return create_new_change_with_params(NOT_ALIVE_DISPOSED, data, wparams, handle);
+}
+
+bool DataWriterImpl::dispose(
+        void* data)
+{
+    logInfo(DATA_WRITER, "Disposing of data");
+    WriteParams wparams;
+    return create_new_change_with_params(NOT_ALIVE_DISPOSED, data, wparams);
+}
+
+bool DataWriterImpl::create_new_change(
+        ChangeKind_t changeKind,
+        void* data)
+{
+    WriteParams wparams;
+    return create_new_change_with_params(changeKind, data, wparams);
+}
+
+
+bool DataWriterImpl::check_new_change_preconditions(
+        ChangeKind_t change_kind,
+        void* data)
+{
+    // Preconditions
+    if (data == nullptr)
+    {
+        logError(PUBLISHER, "Data pointer not valid");
+        return false;
+    }
+
+    if(change_kind == NOT_ALIVE_UNREGISTERED || change_kind == NOT_ALIVE_DISPOSED ||
+            change_kind == NOT_ALIVE_DISPOSED_UNREGISTERED)
+    {
+        if(!type_->m_isGetKeyDefined)
+        {
+            logError(PUBLISHER,"Topic is NO_KEY, operation not permitted");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool DataWriterImpl::perform_create_new_change(
+        ChangeKind_t change_kind,
+        void* data,
+        WriteParams& wparams,
+        const InstanceHandle_t& handle)
+{
+    // Block lowlevel writer
+    auto max_blocking_time = std::chrono::steady_clock::now() +
+        std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.m_reliability.max_blocking_time));
+    std::unique_lock<std::recursive_timed_mutex> lock(writer_->getMutex(), std::defer_lock);
```

Comment:
Insert a line above for code clarity. 
```suggestion

    std::unique_lock<std::recursive_timed_mutex> lock(writer_->getMutex(), std::defer_lock);
```

Reply:
Lock moved.